x-pack/apm-server/sampling/processor.go

```diff
@@ -397,6 +367,7 @@ func (p *Processor) Run() error {
 		case <-ctx.Done():
 			return ctx.Err()
 		case <-p.stopping:
+			<-time.After(shutdownGracePeriod)
```

Comment:
**[For reviewers]** The benchmark reports a high delta for [BenchmarkProcess](https://github.com/elastic/apm-server/blob/cf8f0a616f94aadb00595c81ccdc2a61aa8ebd2d/x-pack/apm-server/sampling/processor_bench_test.go#L22). This is caused due to the introduction of the shutdown grace period.

I have run the same benchmarks by disabling the wait here with the following results (lines with `all equal` are removed):

```
name                                                            old time/op    new time/op    delta
pkg:github.com/elastic/apm-server/internal/agentcfg goos:darwin goarch:arm64
FetchAndAdd/FetchFromCache-10                                     62.1ns ± 2%    62.8ns ± 4%     ~     (p=0.548 n=5+5)
FetchAndAdd/FetchAndAddToCache-10                                  132ns ± 0%     134ns ± 4%   +1.91%  (p=0.032 n=4+5)
pkg:github.com/elastic/apm-server/internal/decoder goos:darwin goarch:arm64
CompressedRequestReader/uncompressed-10                            114ns ± 2%     111ns ± 4%     ~     (p=0.056 n=5+5)
CompressedRequestReader/gzip_content_encoding-10                  4.03µs ± 5%    3.51µs ±12%  -12.78%  (p=0.032 n=4+5)
CompressedRequestReader/gzip_sniff-10                             4.43µs ±65%    3.73µs ±15%     ~     (p=0.841 n=5+5)
CompressedRequestReader/deflate_content_encoding-10               6.49µs ±48%    3.51µs ±23%  -45.99%  (p=0.032 n=5+5)
CompressedRequestReader/deflate_sniff-10                          5.23µs ±43%    3.37µs ± 8%     ~     (p=0.056 n=5+5)
pkg:github.com/elastic/apm-server/internal/model/modelindexer goos:darwin goarch:arm64
ModelIndexer/NoCompression-10                                     1.65µs ±10%    1.54µs ± 6%     ~     (p=0.421 n=5+5)
ModelIndexer/BestSpeed-10                                         3.40µs ±11%    3.22µs ± 0%     ~     (p=0.151 n=5+5)
ModelIndexer/DefaultCompression-10                                5.28µs ± 2%    5.68µs ±21%     ~     (p=0.690 n=5+5)
ModelIndexer/BestCompression-10                                   8.44µs ± 7%    8.38µs ± 8%     ~     (p=0.841 n=5+5)
pkg:github.com/elastic/apm-server/internal/processor/stream goos:darwin goarch:arm64
BackendProcessor-10                                               0.00ns ±59%    0.00ns ±60%     ~     (p=0.286 n=5+5)
RUMV3Processor-10                                                 0.00ns ±48%    0.00ns ±24%     ~     (p=0.095 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel0-10    0.00ns ±67%    0.00ns ±63%     ~     (p=0.841 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel2-10    0.00ns ±21%    0.00ns ±24%  -21.65%  (p=0.032 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel4-10    0.00ns ±25%    0.00ns ±28%     ~     (p=0.111 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel8-10    0.00ns ±36%    0.00ns ±53%     ~     (p=0.222 n=5+5)
pkg:github.com/elastic/apm-server/internal/publish goos:darwin goarch:arm64
Publisher-10                                                       1.00s ± 0%     1.00s ± 0%   +0.16%  (p=0.032 n=5+5)
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/spanmetrics goos:darwin goarch:arm64
AggregateSpan-10                                                   955ns ± 4%    1020ns ±11%     ~     (p=0.381 n=5+5)
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/txmetrics goos:darwin goarch:arm64
AggregateTransaction-10                                            214ns ± 1%     203ns ± 7%     ~     (p=0.095 n=5+5)
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling goos:darwin goarch:arm64
TraceGroups-10                                                     309ns ± 6%     291ns ± 2%   -5.76%  (p=0.016 n=5+5)
Process-10                                                       51.0ms ±135%    76.9ms ±68%     ~     (p=0.548 n=5+5)
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling/eventstorage goos:darwin goarch:arm64
ShardedWriteTransactionUncontended-10                             1.12µs ±28%    1.12µs ±24%     ~     (p=0.730 n=5+5)
ShardedWriteTransactionContended-10                               6.48µs ± 1%    6.46µs ± 1%     ~     (p=0.841 n=5+5)
WriteTransaction/json_codec-10                                    5.63µs ± 2%    5.65µs ± 4%     ~     (p=0.841 n=5+5)
WriteTransaction/json_codec_big_tx-10                             6.79µs ± 3%    6.80µs ± 1%     ~     (p=1.000 n=5+5)
WriteTransaction/nop_codec-10                                      214ns ± 6%     212ns ± 2%     ~     (p=0.841 n=5+5)
WriteTransaction/nop_codec_big_tx-10                               211ns ± 2%     207ns ± 1%   -1.90%  (p=0.040 n=5+5)
ReadEvents/json_codec/0_events-10                                  323ns ± 5%     321ns ± 2%     ~     (p=0.841 n=5+5)
ReadEvents/json_codec/1_events-10                                 11.8µs ±11%    12.1µs ±13%     ~     (p=0.841 n=5+5)
ReadEvents/json_codec/10_events-10                                 103µs ± 2%     101µs ± 0%   -1.95%  (p=0.008 n=5+5)
ReadEvents/json_codec/100_events-10                                942µs ± 3%     927µs ± 2%     ~     (p=0.548 n=5+5)
ReadEvents/json_codec/199_events-10                               1.95ms ± 4%    1.95ms ± 4%     ~     (p=0.841 n=5+5)
ReadEvents/json_codec/399_events-10                               3.96ms ± 1%    3.93ms ± 1%   -0.89%  (p=0.032 n=5+5)
ReadEvents/json_codec/1000_events-10                              10.1ms ± 1%    10.0ms ± 2%     ~     (p=0.548 n=5+5)
ReadEvents/json_codec_big_tx/0_events-10                           327ns ± 2%     322ns ± 2%     ~     (p=0.222 n=5+5)
ReadEvents/json_codec_big_tx/1_events-10                          13.3µs ± 1%    13.9µs ± 4%   +4.29%  (p=0.032 n=5+5)
ReadEvents/json_codec_big_tx/10_events-10                          119µs ± 5%     115µs ± 2%     ~     (p=0.056 n=5+5)
ReadEvents/json_codec_big_tx/100_events-10                        1.12ms ± 3%    1.11ms ± 3%     ~     (p=0.222 n=5+5)
ReadEvents/json_codec_big_tx/199_events-10                        2.24ms ± 3%    2.21ms ± 4%     ~     (p=0.310 n=5+5)
ReadEvents/json_codec_big_tx/399_events-10                        4.60ms ± 4%    4.46ms ± 0%   -3.08%  (p=0.032 n=5+5)
ReadEvents/json_codec_big_tx/1000_events-10                       11.5ms ± 5%    11.4ms ± 2%     ~     (p=0.690 n=5+5)
ReadEvents/nop_codec/0_events-10                                   331ns ± 3%     324ns ± 1%     ~     (p=0.151 n=5+5)
ReadEvents/nop_codec/1_events-10                                  1.99µs ± 5%    1.99µs ± 4%     ~     (p=0.841 n=5+5)
ReadEvents/nop_codec/10_events-10                                 11.0µs ±15%    10.9µs ± 9%     ~     (p=0.841 n=5+5)
ReadEvents/nop_codec/100_events-10                                98.6µs ±17%    92.9µs ±10%     ~     (p=0.310 n=5+5)
ReadEvents/nop_codec/199_events-10                                 179µs ±11%     170µs ± 2%     ~     (p=0.690 n=5+5)
ReadEvents/nop_codec/399_events-10                                 332µs ±20%     326µs ± 4%     ~     (p=0.421 n=5+5)
ReadEvents/nop_codec/1000_events-10                                765µs ±12%     738µs ± 6%     ~     (p=0.690 n=5+5)
ReadEvents/nop_codec_big_tx/0_events-10                            331ns ± 1%     330ns ± 4%     ~     (p=0.254 n=5+5)
ReadEvents/nop_codec_big_tx/1_events-10                           1.99µs ± 7%    1.98µs ± 3%     ~     (p=0.889 n=5+5)
ReadEvents/nop_codec_big_tx/10_events-10                          10.4µs ± 2%    10.6µs ± 1%     ~     (p=0.063 n=5+4)
ReadEvents/nop_codec_big_tx/100_events-10                         94.9µs ±17%    92.6µs ±15%     ~     (p=1.000 n=5+5)
ReadEvents/nop_codec_big_tx/199_events-10                          178µs ±13%     170µs ± 5%     ~     (p=0.841 n=5+5)
ReadEvents/nop_codec_big_tx/399_events-10                          345µs ±20%     342µs ±16%     ~     (p=1.000 n=5+5)
ReadEvents/nop_codec_big_tx/1000_events-10                         826µs ±13%     729µs ±17%     ~     (p=0.095 n=5+5)
IsTraceSampled/sampled-10                                         93.9ns ± 5%    95.5ns ± 4%     ~     (p=0.421 n=5+5)
IsTraceSampled/unsampled-10                                       92.5ns ± 3%    94.6ns ± 4%     ~     (p=0.095 n=5+5)
IsTraceSampled/unknown-10                                          428ns ± 3%     429ns ± 2%     ~     (p=0.690 n=5+5)

name                                                            old alloc/op   new alloc/op   delta
pkg:github.com/elastic/apm-server/internal/agentcfg goos:darwin goarch:arm64
pkg:github.com/elastic/apm-server/internal/decoder goos:darwin goarch:arm64
CompressedRequestReader/gzip_content_encoding-10                  45.5kB ± 0%    45.5kB ± 0%     ~     (p=0.897 n=5+5)
CompressedRequestReader/gzip_sniff-10                             45.5kB ± 0%    45.5kB ± 0%     ~     (p=0.825 n=5+5)
CompressedRequestReader/deflate_content_encoding-10               44.9kB ± 0%    44.9kB ± 0%   -0.05%  (p=0.016 n=5+5)
CompressedRequestReader/deflate_sniff-10                          44.9kB ± 0%    44.9kB ± 0%     ~     (p=0.175 n=5+5)
pkg:github.com/elastic/apm-server/internal/model/modelindexer goos:darwin goarch:arm64
ModelIndexer/NoCompression-10                                     3.28kB ± 2%    3.30kB ± 1%     ~     (p=0.460 n=5+5)
ModelIndexer/BestSpeed-10                                         2.59kB ± 1%    2.58kB ± 0%     ~     (p=0.214 n=5+5)
ModelIndexer/DefaultCompression-10                                2.63kB ± 1%    2.63kB ± 1%     ~     (p=0.690 n=5+5)
ModelIndexer/BestCompression-10                                   2.69kB ± 2%    2.68kB ± 1%     ~     (p=0.643 n=5+5)
pkg:github.com/elastic/apm-server/internal/processor/stream goos:darwin goarch:arm64
pkg:github.com/elastic/apm-server/internal/publish goos:darwin goarch:arm64
Publisher-10                                                       293kB ±11%     279kB ±14%     ~     (p=0.421 n=5+5)
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/spanmetrics goos:darwin goarch:arm64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/txmetrics goos:darwin goarch:arm64
AggregateTransaction-10                                            2.60B ±23%     2.00B ± 0%     ~     (p=0.095 n=5+4)
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling goos:darwin goarch:arm64
Process-10                                                       39.7MB ±124%    59.2MB ±50%     ~     (p=0.421 n=5+5)
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling/eventstorage goos:darwin goarch:arm64
ShardedWriteTransactionUncontended-10                             3.43kB ± 3%    3.54kB ± 4%     ~     (p=0.111 n=5+5)
ShardedWriteTransactionContended-10                               7.63kB ± 0%    7.62kB ± 1%     ~     (p=0.413 n=4+5)
WriteTransaction/json_codec_big_tx-10                             3.52kB ± 0%    3.52kB ± 0%     ~     (p=0.190 n=5+5)
ReadEvents/json_codec/1_events-10                                 8.23kB ± 0%    8.23kB ± 0%     ~     (p=0.714 n=5+5)
ReadEvents/json_codec/10_events-10                                75.3kB ± 0%    75.3kB ± 0%     ~     (p=0.341 n=5+5)
ReadEvents/json_codec/100_events-10                                750kB ± 0%     750kB ± 0%     ~     (p=0.690 n=5+5)
ReadEvents/json_codec/199_events-10                               1.02MB ± 0%    1.02MB ± 0%   -0.09%  (p=0.040 n=5+5)
ReadEvents/json_codec/399_events-10                               1.57MB ± 0%    1.56MB ± 0%     ~     (p=0.270 n=5+5)
ReadEvents/json_codec/1000_events-10                              3.36MB ± 0%    3.29MB ± 0%   -2.27%  (p=0.029 n=4+4)
ReadEvents/json_codec_big_tx/1_events-10                          10.1kB ± 0%    10.1kB ± 0%     ~     (p=0.643 n=5+5)
ReadEvents/json_codec_big_tx/10_events-10                         94.5kB ± 0%    94.5kB ± 0%     ~     (p=0.690 n=5+5)
ReadEvents/json_codec_big_tx/100_events-10                         942kB ± 0%     942kB ± 0%     ~     (p=0.135 n=5+5)
ReadEvents/json_codec_big_tx/199_events-10                        1.33MB ± 0%    1.33MB ± 0%     ~     (p=0.571 n=5+5)
ReadEvents/json_codec_big_tx/399_events-10                        2.11MB ± 0%    2.11MB ± 0%     ~     (p=0.524 n=5+4)
ReadEvents/json_codec_big_tx/1000_events-10                       4.62MB ± 0%    4.62MB ± 0%     ~     (p=0.571 n=5+4)
ReadEvents/nop_codec/1_events-10                                  3.25kB ± 0%    3.26kB ± 0%     ~     (p=0.381 n=5+5)
ReadEvents/nop_codec/10_events-10                                 25.6kB ± 0%    25.6kB ± 0%     ~     (p=1.000 n=5+5)
ReadEvents/nop_codec/100_events-10                                 249kB ± 0%     249kB ± 0%     ~     (p=0.310 n=5+5)
ReadEvents/nop_codec/199_events-10                                 467kB ± 0%     467kB ± 0%     ~     (p=0.310 n=5+5)
ReadEvents/nop_codec/399_events-10                                 890kB ± 0%     890kB ± 0%     ~     (p=1.000 n=5+5)
ReadEvents/nop_codec/1000_events-10                               2.15MB ± 0%    2.15MB ± 0%     ~     (p=0.230 n=5+5)
ReadEvents/nop_codec_big_tx/1_events-10                           3.25kB ± 0%    3.25kB ± 0%     ~     (p=1.000 n=5+4)
ReadEvents/nop_codec_big_tx/10_events-10                          25.6kB ± 0%    25.6kB ± 0%     ~     (p=0.087 n=5+5)
ReadEvents/nop_codec_big_tx/100_events-10                          249kB ± 0%     249kB ± 0%     ~     (p=0.690 n=5+5)
ReadEvents/nop_codec_big_tx/199_events-10                          467kB ± 0%     467kB ± 0%     ~     (p=0.302 n=5+4)
ReadEvents/nop_codec_big_tx/399_events-10                          890kB ± 0%     889kB ± 0%     ~     (p=0.516 n=5+5)
ReadEvents/nop_codec_big_tx/1000_events-10                        2.15MB ± 0%    2.14MB ± 0%   -0.33%  (p=0.008 n=5+5)
IsTraceSampled/unknown-10                                           334B ± 4%      334B ± 4%     ~     (p=1.000 n=5+5)

name                                                            old allocs/op  new allocs/op  delta
pkg:github.com/elastic/apm-server/internal/agentcfg goos:darwin goarch:arm64
pkg:github.com/elastic/apm-server/internal/decoder goos:darwin goarch:arm64
pkg:github.com/elastic/apm-server/internal/model/modelindexer goos:darwin goarch:arm64
pkg:github.com/elastic/apm-server/internal/processor/stream goos:darwin goarch:arm64
pkg:github.com/elastic/apm-server/internal/publish goos:darwin goarch:arm64
Publisher-10                                                       1.21k ±14%     1.21k ±14%     ~     (p=1.000 n=5+5)
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/spanmetrics goos:darwin goarch:arm64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/txmetrics goos:darwin goarch:arm64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling goos:darwin goarch:arm64
Process-10                                                          467 ±120%       681 ±50%     ~     (p=0.548 n=5+5)
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling/eventstorage goos:darwin goarch:arm64
ReadEvents/json_codec/1000_events-10                               12.5k ± 0%     12.5k ± 0%     ~     (p=0.333 n=4+5)
```

I think, having a fixed grace period for shutdown should be okay for our case since it only comes into the picture while stopping the processor, additionally, during reload, we first start a new processor and then stop the older one. Let me know if my assumptions are not correct here, I think we can invest some time in having hooks to properly detect clean stop of the sampling processor rather than just waiting for a fixed amount of time.

Reply:
Since the grace period was added in `Processor#Stop`, moving `Processor#Stop` to `Cleanup` removes the added time from benchmark results ([relevant commit](https://github.com/elastic/apm-server/pull/8809/commits/4e092dc5036e4c70c2967b329ecbeb0f5de2b2db))